src/ui/EditorNS/editor.cpp

```diff
@@ -268,117 +305,184 @@ namespace EditorNS
+            newMode.name = mode.value("name").toString();
+            newMode.mime = mode.value("mime").toString();
+            newMode.mode = mode.value("mode").toString();
+            newMode.fileNames = mode.value("fileNames").toVariant().toStringList();
+            newMode.fileExtensions = mode.value("fileExtensions").toVariant().toStringList();
+            newMode.firstNonBlankLine = mode.value("firstNonBlankLine").toVariant().toStringList();
+
+            m_langCache.append(newMode);
         }
+    }
 
-        QMap<QString, QVariant> languages =
-            engine.evaluate("Languages.languages", fileName).toVariant().toMap();
-
-        QList<QMap<QString, QString>> langs;
-
-        QMap<QString, QVariant>::iterator lang;
-        for (lang = languages.begin(); lang != languages.end(); ++lang) {
-            QMap<QString, QVariant> mode = lang.value().toMap();
-
-            QMap<QString, QString> newMode;
-            newMode.insert("id", lang.key());
-            newMode.insert("name", mode.value("name").toString());
-            newMode.insert("mime", mode.value("mime").toString());
-            newMode.insert("mode", mode.value("mode").toString());
-            langs.append(newMode);
+    QVariant Editor::getLanguageVariantData(const QString& langId)
+    {
+        auto comp = QVariant::fromValue(langId);
+        for (const auto& l : m_langCache) {
+            if (l.id == langId) {
+                QVariantMap langData;
+                langData.insert("id", l.id);
+                langData.insert("mime", l.mime);
+                langData.insert("mode", l.mode);
+                langData.insert("name", l.name);
+                langData.insert("fileNames", l.fileNames);
+                langData.insert("fileExtensions", l.fileExtensions);
+                return QVariant::fromValue(langData);
+            }
         }
-        return langs;
+        return QVariant();
+    }
+
+    QVector<Editor::LanguageData> Editor::languages()
+    {
+        return m_langCache;
     }
 
-    QString Editor::getLanguage()
+    Editor::LanguageData Editor::getLanguage()
     {
-        QVariantMap data = m_jsProxy->getRawValue("language").toMap();
-        return data.value("id").toString();
+        return m_info.language;
     }
 
     void Editor::setLanguage(const QString &language)
     {
-        sendMessage("C_CMD_SET_LANGUAGE", language);
-        if (!m_customIndentationMode)
+        for (const auto& l : m_langCache) {
+            if (l.id == language) {
+                m_info.language = l;
+                break;
+            }
+        }
+        if (!m_info.content.indentMode.custom) {
             setIndentationMode(language);
+        }
+        QString mode;
+        if (m_info.language.mime.isEmpty()) {
+            mode = m_info.language.mode;
+        } else {
+            mode = m_info.language.mime;
+        }
+        sendMessage("C_CMD_SET_LANGUAGE", mode);
+        emit languageChanged(m_info.language);
```

Comment:
Actually, scratch that... the editor.on("option") event in the new CodeMirror will solve this issue as well.